File: android/app/src/main/java/com/videoconverter/android/domain/Concat.kt
package com.videoconverter.android.domain

import java.util.Locale

const val VIDEO_CONCAT_PRESET_ID = "video-concat"
const val VIDEO_CONCAT_MAX_SOURCES = 20

data class ConcatTarget(
    val width: Int,
    val height: Int,
    val frameRate: Double,
)

fun isVideoConcatPreset(preset: String): Boolean = preset == VIDEO_CONCAT_PRESET_ID

fun concatOutputStem(displayName: String): String = "${sourceStem(displayName)}-merged"

fun evenDimension(value: Int): Int = maxOf(2, value - (value % 2))

fun concatTarget(
    first: MediaInfo,
    missingVideo: String = "Each clip needs a video track",
): ConcatTarget {
    val width = first.width
    val height = first.height
    if (width == null || height == null || width <= 0 || height <= 0) {
        throw IllegalArgumentException(missingVideo)
    }
    return ConcatTarget(
        width = evenDimension(width),
        height = evenDimension(height),
        frameRate = first.frameRate ?: 30.0,
    )
}

fun concatScalePadFilter(target: ConcatTarget): String {
    val w = target.width
    val h = target.height
    val fps = concatSecs(target.frameRate)
    return "scale=$w:$h:force_original_aspect_ratio=decrease,pad=$w:$h:(ow-iw)/2:(oh-ih)/2,setsar=1,fps=$fps"
}

fun buildConcatNormalizeArgs(
    input: String,
    outputPartial: String,
    media: MediaInfo,
    target: ConcatTarget,
    quality: String,
    preferHardware: Boolean = true,
): List<String> {
    val args = mutableListOf(
        "-hide_banner",
        "-nostats",
        "-progress",
        "pipe:1",
        "-y",
        "-i",
        ffmpegFileArg(input),
    )
    val duration = maxOf(media.durationSecs ?: 0.0, 0.05)
    val missingAudio = media.audioCodec == null
    if (missingAudio) {
        args += listOf(
            "-f",
            "lavfi",
            "-t",
            concatSecs(duration),
            "-i",
            "anullsrc=channel_layout=stereo:sample_rate=48000",
        )
    }
    args += listOf("-c:v", ffmpegVideoCodec("h264", preferHardware))
    args += if (preferHardware) {
        hardwareVideoQualityArgs(quality)
    } else {
        softwareVideoQualityArgs("h264", quality)
    }
    args += listOf(
        "-pix_fmt",
        "yuv420p",
        "-vf",
        concatScalePadFilter(target),
    )
    if (missingAudio) {
        args += listOf("-map", "0:v:0", "-map", "1:a:0", "-shortest")
    }
    args += listOf(
        "-c:a",
        "aac",
        "-ar",
        "48000",
        "-ac",
        "2",
        "-b:a",
        "192k",
        "-movflags",
        "+faststart",
        "-f",
        "mp4",
        ffmpegFileArg(outputPartial),
    )
    return args
}

fun buildConcatJoinArgs(listPath: String, outputPartial: String): List<String> = listOf(
    "-hide_banner",
    "-nostats",
    "-progress",
    "pipe:1",
    "-y",
    "-f",
    "concat",
    "-safe",
    "0",
    "-i",
    ffmpegFileArg(listPath),
    "-c",
    "copy",
    "-movflags",
    "+faststart",
    "-f",
    "mp4",
    ffmpegFileArg(outputPartial),
)

fun concatListFileContents(paths: List<String>): String =
    paths.joinToString("\n") { path ->
        val escaped = path.replace("'", "'\\''")
        "file '$escaped'"
    }

private fun concatSecs(value: Double): String = "%.3f".format(Locale.US, value)

File: ios/LiteTrans/Domain/Concat.swift
import Foundation

public struct ConcatTarget: Equatable, Sendable {
    public var width: Int
    public var height: Int
    public var frameRate: Double
    public init(width: Int, height: Int, frameRate: Double) {
        self.width = width
        self.height = height
        self.frameRate = frameRate
    }
}

public func concatOutputStem(_ displayName: String) -> String {
    "\(sourceStem(displayName))-merged"
}

public func evenDimension(_ value: Int) -> Int {
    max(2, value - (value % 2))
}

public func concatTarget(from first: MediaInfo) throws -> ConcatTarget {
    guard let width = first.width, let height = first.height, width > 0, height > 0 else {
        throw LiteTransError.concatMissingVideo
    }
    return ConcatTarget(
        width: evenDimension(width),
        height: evenDimension(height),
        frameRate: first.frameRate ?? 30
    )
}

public func concatScalePadFilter(target: ConcatTarget) -> String {
    let w = target.width
    let h = target.height
    let fps = String(format: "%.3f", locale: Locale(identifier: "en_US_POSIX"), target.frameRate)
    return "scale=\(w):\(h):force_original_aspect_ratio=decrease,pad=\(w):\(h):(ow-iw)/2:(oh-ih)/2,setsar=1,fps=\(fps)"
}

public func buildConcatNormalizeArgs(
    input: String,
    outputPartial: String,
    media: MediaInfo,
    target: ConcatTarget,
    quality: String,
    preferHardware: Bool = true
) -> [String] {
    var args = [
        "-hide_banner", "-nostats", "-progress", "pipe:1", "-y",
        "-i", ffmpegFileArg(input),
    ]
    let duration = max(media.durationSecs ?? 0, 0.05)
    let missingAudio = media.audioCodec == nil
    if missingAudio {
        args += [
            "-f", "lavfi",
            "-t", String(format: "%.3f", locale: Locale(identifier: "en_US_POSIX"), duration),
            "-i", "anullsrc=channel_layout=stereo:sample_rate=48000",
        ]
    }
    args += ["-c:v", ffmpegVideoCodec("h264", preferHardware: preferHardware)]
    if preferHardware {
        args += hardwareVideoQualityArgs(quality)
    } else {
        args += softwareVideoQualityArgs(encoder: "h264", quality: quality)
    }
    args += [
        "-pix_fmt", "yuv420p",
        "-vf", concatScalePadFilter(target: target),
    ]
    if missingAudio {
        args += ["-map", "0:v:0", "-map", "1:a:0", "-shortest"]
    }
    args += [
        "-c:a", "aac", "-ar", "48000", "-ac", "2", "-b:a", "192k",
        "-movflags", "+faststart",
        "-f", "mp4",
        ffmpegFileArg(outputPartial),
    ]
    return args
}

public func buildConcatJoinArgs(listPath: String, outputPartial: String) -> [String] {
    [
        "-hide_banner", "-nostats", "-progress", "pipe:1", "-y",
        "-f", "concat", "-safe", "0",
        "-i", ffmpegFileArg(listPath),
        "-c", "copy",
        "-movflags", "+faststart",
        "-f", "mp4",
        ffmpegFileArg(outputPartial),
    ]
}

public func concatListFileContents(paths: [String]) -> String {
    paths.map { path in
        let escaped = path.replacingOccurrences(of: "'", with: "'\\''")
        return "file '\(escaped)'"
    }.joined(separator: "\n")
}
